src/oni2-keyboard-layout/stubs/keyboard-layout-mac.c

```diff
@@ -0,0 +1,240 @@
+  }
+
+  caml_release_runtime_system();
+}
+
+CAMLprim value oni2_KeyboardLayoutInit() {
+  CAMLparam0();
+
+  CFNotificationCenterAddObserver(
+    CFNotificationCenterGetDistributedCenter(),
+    NULL,
+    notificationHandler,
+    kTISNotifySelectedKeyboardInputSourceChanged,
+    NULL,
+    CFNotificationSuspensionBehaviorDeliverImmediately
+  );
+
+  CAMLreturn(Val_unit);
+}
+
+CAMLprim value oni2_KeyboardLayoutGetCurrentLayout() {
+  CAMLparam0();
+  CAMLlocal1(vLayout);
+
+  vLayout = caml_copy_string([oni2_priv_GetCurrentKeyboardLayout() UTF8String]);
+
+  CAMLreturn(vLayout);
+}
+
+CAMLprim value oni2_KeyboardLayoutGetCurrentLanguage() {
+  CAMLparam0();
+  CAMLlocal1(vLanguage);
+
+  vLanguage = caml_copy_string([oni2_priv_GetCurrentKeyboardLanguage() UTF8String]);
+
+  CAMLreturn(vLanguage);
+}
+
+/* Given a keyboard layout, a key code, and modifiers,
+   fill a buffer of UTF-16 characters with the resulting
+   string
+ */
+size_t characterForNativeCode(
+  const UCKeyboardLayout *keyboardLayout,
+  uint16_t virtualKeyCode,
+  EventModifiers modifiers, 
+  unichar *characters
+) {
+  uint32_t modifierKeyState = (modifiers >> 8) & 0xFF;
+  uint32_t deadKeyState = 0;
+  size_t charCount = 0;
+
+  OSStatus status = UCKeyTranslate(
+    keyboardLayout,
+    (uint16_t)virtualKeyCode,
+    kUCKeyActionDown,
+    modifierKeyState,
+    LMGetKbdLast(),
+    kUCKeyTranslateNoDeadKeysBit,
+    &deadKeyState,
+    4 * sizeof(characters[0]),
+    &charCount,
+    characters
+  );
+
+  if (status == noErr && deadKeyState != 0) {
+    status = UCKeyTranslate(
+      keyboardLayout,
+      (uint16_t)virtualKeyCode,
+      kUCKeyActionDown,
+      modifierKeyState,
+      LMGetKbdLast(),
+      kUCKeyTranslateNoDeadKeysBit,
+      &deadKeyState,
+      4 * sizeof(characters[0]),
+      &charCount,
+      characters
+    );
+  }
+
+  if (status == noErr && !iscntrl(characters[0])) {
+    return charCount;
+  } else {
+    return 0;
+  }
+}
+
+CAMLprim value oni2_KeyboardLayoutPopulateCurrentKeymap(value keymap, value Hashtbl_replace) {
+  CAMLparam2(keymap, Hashtbl_replace);
+  CAMLlocal2(keymapEntry, vDom3Code);
+
+  // Allocate string pointers to be filled
+  const char *unmodified;
+  const char *withShift;
+  const char *withAltGraph;
+  const char *withAltGraphShift;
+
+  // Allocate UTF-16 Buffers to be populated
+  unichar uniUnmodified[4]; 
+  unichar uniWithShift[4]; 
+  unichar uniWithAltGraph[4]; 
+  unichar uniWithAltGraphShift[4]; 
+
+  TISInputSourceRef source = TISCopyCurrentKeyboardInputSource();
+  CFDataRef layoutData = (CFDataRef)TISGetInputSourceProperty(source, kTISPropertyUnicodeKeyLayoutData);
+
+  if (layoutData == NULL) {
+    CAMLreturn(Val_unit);
+  }
+
+  const UCKeyboardLayout *keyboardLayout = (UCKeyboardLayout *)CFDataGetBytePtr(layoutData);
+
+  size_t keyCodeMapSize = sizeof(keyCodeMap) / sizeof(keyCodeMap[0]);
+  for (size_t i = 0; i < keyCodeMapSize; i++) {
+    const char *dom3Code = keyCodeMap[i].dom3Code;
```

Comment:
One thing I need to explore is how to map the SDL keys to the keymap here - I'm not sure there actually are values corresponding to the `dom3Code`.

So I'll need to either:
1) Have a function that maps from SDL key <-> dom3code
2) Use another value available on the key (either keycode or scancode) to map to these entries

Reply:
That's a good point. I looked into this and I couldn't find if there's a way to convert native scancodes/keycodes to SDL2 ones. The internal representation of SDL's scancodes/keycodes seem to be decoupled from the raw USB codes that `keyboard-layout` uses. This means we would have to write a mapping either way -- either `SDL2 Keycode => DOM3 Keycode` or `SDL2 Keycode => Native Keycode`. I think the first option is easier since we wouldn't have to write a mapping for each OS since they all have different keycodes:
https://github.com/onivim/oni2/blob/dc53ea4a3075d7ea267b0ec509d4c2581be4bed8/src/oni2-keyboard-layout/stubs/keycode_converter_data.h#L125-L133

Reply:
I actually realized we could add the mapping in that header file, so in 384d465 I painstakingly added all the SDL2 scancodes and keycodes to the mapping. That should let us construct a `Hashtbl.t` that maps `SDL2 Scancode => Keymap Entry`

Reply:
> That should let us construct a `Hashtbl.t` that maps `SDL2 Scancode => Keymap Entry`

I did this in 0787940 -- let me know what you think! The tests now use the new API:
https://github.com/onivim/oni2/blob/0787940af366828aea4c381f9180454debb0e46a/test/oni2-keyboard-layout/KeymapTest.re#L14-L15